### PR TITLE
Add unit tests to shared util functions with no dependencies

### PR DESCRIPTION
**Pull request by leoraba**
Changes:
- Installed testing tools (mocha and chai) to run Test suites
- Adding all tests on `test/` folder 
- Unit Testing on `utils` that doesn't need any external dependency.
- Total Coverage 8.3% of lines

Fix issue #73

### CODE REVIEW

File: packages/data-provider/test/utils/convertSqonToQuery.spec.ts [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+import { describe, it } from 'mocha';
+
+import { SQON } from '@overture-stack/sqon-builder';
+
+import { convertSqonToQuery, parseSQON } from '../../src/utils/convertSqonToQuery.js';
+
+function extractValues(obj: SQL | undefined, key: string) {
+	let values: SQL[] = [];
+
+	function recurse(currentObj: SQL | undefined) {
```
Comment (justincorrigible) [inline packages/data-provider/test/utils/convertSqonToQuery.spec.ts:12]: this function needn't be defined withing extractValues, as far as I can tell.

File: packages/data-provider/test/utils/dictionarySchemasTestData.ts [hunk 2]
```diff
@@ -0,0 +1,264 @@
+import { Dictionary } from '@overture-stack/lyric-data-model';
+import { SchemaDefinition, ValueType } from '@overturebio-stack/lectern-client/lib/schema-entities.js';
+
+const sampleDictionaryData: SchemaDefinition[] = [
+	{
+		name: 'study',
```
Comment (justincorrigible) [inline packages/data-provider/test/utils/dictionarySchemasTestData.ts:6]: is it reasonable to suggest we use a more generic test model? study and participant seem PCGL-specific, when we could be using fruits or animals to clearly exemplify the reusability of the software.

**Comment by justincorrigible** [comment]
happy to see all this testing 💪 
great work, @leoraba!

File: packages/data-provider/test/utils/convertSqonToQuery.spec.ts (hunk 1 above)
Comment (leoraba) [reply packages/data-provider/test/utils/convertSqonToQuery.spec.ts:12]: Added types on this function as mocha tries to do typescript checks and couldn't find a way to exclude  on test files.

The purpose of this function is to convert the input SQL object from drizzle into an Array which make it easier to compare on `expect()` function

Comment (justincorrigible) [reply packages/data-provider/test/utils/convertSqonToQuery.spec.ts:18]: to be clear, what I mean is lines 18-28 could be cut and pasted to line 8 (outside `extractValues`) and the result would be the same, minus the reinstantiation of `recurse` every time the parent function runs

File: packages/data-provider/test/utils/convertSqonToQuery.spec.ts [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+			op: 'gt',
+			content: { fieldName: 'date_of_birth', value: 197005 },
+		};
+
+		const greaterThanFilterChunk: SQLChunk[] = ["data ->> 'date_of_birth' > '197005'"];
+
+		it('should convert a json text with greater than filter into a SQON format', () => {
+			const result = parseSQON(sqonGreaterThanFilterRawInput);
+			expect(JSON.stringify(result)).to.eql(JSON.stringify(sqonGreaterThanFilterParsed));
+		});
```
Comment (joneubank) [inline packages/data-provider/test/utils/convertSqonToQuery.spec.ts:59]: `parseSQON` appears to just be a wrapper over the SQONBuilder library... it doesn't really need test cases. If there are specific requirements you need this to perform then certainly write those. This test just checks if the outer `and` combination operator is stripped from the SQON...

File: packages/data-provider/test/utils/convertSqonToQuery.spec.ts [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+		it('should convert SQON with greater than filter into a database query', () => {
+			const result = convertSqonToQuery(sqonGreaterThanFilterParsed);
+			const extractedValues = extractValues(result, 'value');
+			expect(extractedValues).to.eql(greaterThanFilterChunk);
+		});
```
Comment (joneubank) [inline packages/data-provider/test/utils/convertSqonToQuery.spec.ts:65]: Good test.

File: packages/data-provider/test/utils/convertSqonToQuery.spec.ts [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+			op: 'lt',
+			content: { fieldName: 'date_of_birth', value: 197005 },
+		};
+
+		const lessThanFilterChunk: SQLChunk[] = ["data ->> 'date_of_birth' < '197005'"];
+
+		it('should convert a json text with less than filter into a SQON format', () => {
+			const result = parseSQON(sqonLessThanFilterRawInput);
+			expect(JSON.stringify(result)).to.eql(JSON.stringify(sqonLessThanFilterParsed));
+		});
```
Comment (joneubank) [inline packages/data-provider/test/utils/convertSqonToQuery.spec.ts:92]: Repeat comment about testing parseSQON.

File: packages/data-provider/test/utils/dictionarySchemasTestData.ts [hunk 6]
```diff
@@ -0,0 +1,213 @@
+import { Dictionary } from '@overture-stack/lyric-data-model';
+import { SchemaDefinition, ValueType } from '@overturebio-stack/lectern-client/lib/schema-entities.js';
+
+const sampleDictionaryData: SchemaDefinition[] = [
```
Comment (joneubank) [inline packages/data-provider/test/utils/dictionarySchemasTestData.ts:4]: You may want to create a dir of `test/fixtures` that includes test dictionaries

Comment (joneubank) [reply packages/data-provider/test/utils/dictionarySchemasTestData.ts:4]: This dictionary could be called `dictionarySportStats`

File: packages/data-provider/test/utils/dictionarySchemasTestData.ts [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+export const sampleDictionary: Dictionary = {
+	id: 1,
+	name: 'test dictionary',
+	version: '1.0.0',
+	dictionary: sampleDictionaryData,
+	createdAt: new Date(),
+	createdBy: '',
+};
+
+export const sampleChildrenNodes = {
```
Comment (joneubank) [inline packages/data-provider/test/utils/dictionarySchemasTestData.ts:201]: `dictionarySportStatsNodeGraph` or something more descriptive. We will end up with many `samples` as we develop.

File: packages/data-provider/test/utils/dictionarySchemasTestData.ts [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+		{ schemaName: 'game', fieldName: 'sport_id', parent: { schemaName: 'sport', fieldName: 'sport_id' } },
+	],
+	team: [
+		{
+			schemaName: 'player',
+			fieldName: 'team_id',
+			parent: { schemaName: 'team', fieldName: 'team_id' },
+		},
+	],
+};
```
Comment (joneubank) [inline packages/data-provider/test/utils/dictionarySchemasTestData.ts:213]: `as const` so no tests try to modify our test fixture.

Comment (joneubank) [reply packages/data-provider/test/utils/dictionarySchemasTestData.ts:213]: Repeat this comment for all the test fixtures.

File: packages/data-provider/test/utils/fileUtils.spec.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
+import { expect } from 'chai';
+import fs from 'fs';
+import { describe, it } from 'mocha';
+import sinon from 'sinon';
+
+const fsPromises = fs.promises;
+
+import { tsvToJson } from '../../src/utils/fileUtils.js';
+
+describe('File Utils tests', () => {
```
Comment (joneubank) [inline packages/data-provider/test/utils/fileUtils.spec.ts:10]: are there any failure cases we should test? behaviour we expect when the file is not a tsv? empty file behaviour? your example only has a single record, should we test one with multiple records? a header row and no records?

File: packages/data-provider/test/utils/formatUtils.spec.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+describe('Format Utils tests', () => {
+	describe('Validate if input is a valid ID number', () => {
+		it('should return true if input is a valid number', () => {
+			const validNumber = 1;
+			const response = isValidIdNumber(validNumber);
+			expect(response).to.be.true;
+		});
+
+		it('should return false if input is a not a valid number', () => {
+			const invalidNumber = 'one';
```
Comment (joneubank) [inline packages/data-provider/test/utils/formatUtils.spec.ts:15]: Are there other edge cases to test? empty values, numeric strings like `'1'`, `Infinity`/`NaN`, null/undefined, negative numbers, `0`, numbers beyond max int...

File: packages/data-provider/test/utils/formatUtils.spec.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+			const response = isValidIdNumber(invalidNumber);
+			expect(response).to.be.false;
+		});
+	});
+
+	describe('Extract unique characters on a String', () => {
+		it('should return unique characters on a string', () => {
+			const duplicateCharactersString = 'AAABBC';
+			const uniqueCharactersString = 'ABC';
+			const response = uniqueCharacters(duplicateCharactersString);
```
Comment (joneubank) [inline packages/data-provider/test/utils/formatUtils.spec.ts:25]: What do we use this for? Seems cool...

File: packages/data-provider/test/utils/formatUtils.spec.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+			expect(response).to.eql(uniqueCharactersString);
+		});
+	});
+
+	describe('Validate if input string is empty or not', () => {
+		it('should return true if string is empty', () => {
+			const emptyString = '';
```
Comment (joneubank) [inline packages/data-provider/test/utils/formatUtils.spec.ts:32]: What about the string `'       '`

File: packages/data-provider/test/utils/formatUtils.spec.ts [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+			const response = isEmptyString(emptyString);
+			expect(response).to.be.true;
+		});
+
+		it('should return false if string is not empty', () => {
+			const emptyString = 'test';
```
Comment (joneubank) [inline packages/data-provider/test/utils/formatUtils.spec.ts:38]: other edge case? null/undefined/false/0/NaN/[]

File: packages/data-provider/test/utils/convertSqonToQuery.spec.ts [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+			}
+			Object.values(currentObj).forEach(recurse);
+		}
+	}
+
+	recurse(obj);
+	return values;
+}
+
+describe('Test SQON functions', () => {
```
[the remark this answers is not in the document]
Comment (leoraba) [reply packages/data-provider/test/utils/convertSqonToQuery.spec.ts:34]: renamed as suggested

File: packages/data-provider/test/utils/convertSqonToQuery.spec.ts (hunk 3 above)
Comment (leoraba) [reply packages/data-provider/test/utils/convertSqonToQuery.spec.ts:59]: `parseSQON` is a wrapper of a SQONBuilder function and we're not adding any logic. Will remove this test .

File: packages/data-provider/test/utils/dictionarySchemasTestData.ts (hunk 6 above)
Comment (leoraba) [reply packages/data-provider/test/utils/dictionarySchemasTestData.ts:4]: Moved this file to `fixtures/` folder and changed it's variable names

File: packages/data-provider/test/utils/dictionarySchemasTestData.ts (hunk 7 above)
Comment (leoraba) [reply packages/data-provider/test/utils/dictionarySchemasTestData.ts:201]: renamed const name to `dictionarySportStatsNodeGraph `

File: packages/data-provider/test/utils/dictionarySchemasTestData.ts (hunk 8 above)
Comment (leoraba) [reply packages/data-provider/test/utils/dictionarySchemasTestData.ts:213]: Good catch. Added const assertion.

File: packages/data-provider/test/utils/formatUtils.spec.ts (hunk 11 above)
Comment (leoraba) [reply packages/data-provider/test/utils/formatUtils.spec.ts:25]: we use this function to generate system IDs using only unique characters when user sets a custom alphabet

File: packages/data-provider/test/utils/fileUtils.spec.ts (hunk 9 above)
Comment (leoraba) [reply packages/data-provider/test/utils/fileUtils.spec.ts]: adding more tests to cover more edge scenarios

File: README.md [hunk 15]
```diff
@@ -26,7 +26,7 @@ The structure of this monorepo is app centric having `apps/` folder to keep depl
 
 - [PNPM](https://pnpm.io/) Project manager
 - [Node.js](https://nodejs.org/en) Runtime environment (v20 or higher)
-- [VS Code](https://code.visualstudio.com/) As recommended code editor
+- [VS Code](https://code.visualstudio.com/) As recommended code editor. Plugins recommended: ESLint, Prettier - Code formatter, Mocha Test Explorer, Monorepo Workspace
```
Comment (joneubank) [inline README.md:29]: 👍

**Review by joneubank — APPROVED** [review]
Great stuff, thanks.
